### ISSUE TITLE
`TimeoutError: Promise timed out after 500 milliseconds` on client disconnect

### ISSUE THREAD

**Issue by addisonElliott** [OptimalBits/bull#2302]
## Description

I'll apologize in advance for not having a simple, reproducible case. If I get some time in the future, I'll work on creating a test case that fails.

I'm running tests using jest and they started failing when I upgraded Bull. The bug starts in v4.5.5. Downgrading to v4.5.4, the problem is no longer present. See below for the error message.

```
  ● Test suite failed to run

    TimeoutError: Promise timed out after 500 milliseconds

      at Timeout.<anonymous> (node_modules/bull/lib/p-timeout.js:56:50)
```

This is the relevant change that caused issues: https://github.com/OptimalBits/bull/commit/41b940457b3447619c3c2887674a8cebf1508b07

The old `p-timeout` version used in v4.5.4 is v3.2.0 (https://github.com/sindresorhus/p-timeout/blob/v3.2.0/index.js). If you compare that with the latest version (which is what was upgraded to), it's pretty different. I suspect the bug lies somewhere in there.

That's about as far as I got ☹️. I spent some time debugging it but didn't get anywhere useful.

Let me know how I can help

## Bull version

4.5.5

**Comment by pauldeng** [comment]
+1

I am experiencing this issue as well. Unfortunately, I cannot add more detail to this ticket.

**Comment by manast** [comment]
@addisonElliott when you say "starts failing", do you mean that some of your test cases are failing now with the above-mentioned error? if so it would be great if you can strip down such case that reproduces it. I will keep this issue open waiting for such a case anyway.

**Comment by cartuchogl** [comment]
I have observed the same problem, when I close my workers from the console with Ctrl-C. With 4.5.4 as @addisonElliott points out it doesn't happen. As the change 41b9404 also mentions as the cause of the problem, I have ventured to change the lib/queue.js on line 545 from finally to catch and the problem disappears. But I suspect that it must not be correct.

Thank you for your support.

**Comment by manast** [comment]
Replacing to catch will leave the queue hanging in some cases, I do not recommend you to do that. If you provide a reproducible case then I can look into it quickly.

**Comment by Jhoanegar** [comment]
Nice detective work @[addisonElliott](https://github.com/addisonElliott)!  I also experienced this issue while trying to update from `3.20.1` to the latest. 

I just wanted to mention for anyone else out there that this error is particularly hard to find if you are tearing down your queues inside a docker container, for me, the container simply froze and never failed. After upgrading docker desktop on my mac to the latest and deleting all the cached volumes and images I started seeing the same error message and that finally led me to this thread, downgrading like you did fixed the issue for me.

**Comment by cartuchogl** [comment]
@manast thanks for your answer, I am now using 4.5.4.
I said it in case it helped.

**Comment by addisonElliott** [comment]
> @addisonElliott when you say "starts failing", do you mean that some of your test cases are failing now with the above-mentioned error? if so it would be great if you can strip down such case that reproduces it. I will keep this issue open waiting for such a case anyway.

Yep, on 4.5.4 my tests work and 4.5.5 they fail. More specifically, the tests will pass but the test suite fails (presumably because of a hanging promise or something?).

## Test case

Here's a simple test case that reproduces my issue. Notably, it appears that the `obliterate()` calls have something to do with it...

https://gist.github.com/addisonElliott/b16053072327587ad3562e35b61026a6

**Comment by addisonElliott** [comment]
I just manually applied your fix in #2303 and that **does** fix my issue. It looks like it causes other tests to fail.

Once again, I spent some time debugging this but quickly got lost in the weeds. I thought the problem was due to the pTimeout promise not being cleared.
